src/io/video.js

```diff
@@ -0,0 +1,332 @@
+     */
+    setPreviewGhost (ghost) {
+        if (this._drawable) {
+            this.runtime.renderer.updateDrawableProperties(this._drawable, {ghost});
+        }
+    }
+
+
+    /**
+     * Create a video stream.
+     * Should probably be moved to -render or somewhere similar later
+     * @private
+     * @return {Promise} When video has been received, rejected if video is not received
+     */
+    _setupVideo () {
+        if (this._singleSetup) {
+            return this._singleSetup;
+        }
+
+        this._video = document.createElement('video');
+        const video = new Promise((resolve, reject) => {
+            navigator.getUserMedia({
+                audio: false,
+                video: {
+                    width: {min: 480, ideal: 640},
+                    height: {min: 360, ideal: 480}
+                }
+            }, stream => {
+                this._video.src = window.URL.createObjectURL(stream);
+                // Hint to the stream that it should load. A standard way to do this
+                // is add the video tag to the DOM. Since this extension wants to
+                // hide the video tag and instead render a sample of the stream into
+                // the webgl rendered Scratch canvas, another hint like this one is
+                // needed.
+                this._track = stream.getTracks()[0];
+                resolve(this._video);
+            }, err => {
+                // There are probably some error types we could handle gracefully here.
+                this._singleSetup = null;
+                reject(err);
+            });
+        });
+
+        this._singleSetup = video.then(() => this._setupPreview());
+        return this._singleSetup;
+    }
+
+    _disablePreview () {
+        if (this._skin && this._drawable) {
+            this._skin.clear();
+            this.runtime.renderer.updateDrawableProperties(this._drawable, {visible: false});
+        }
+        this._renderPreviewFrame = null;
+    }
+
+    _setupPreview () {
+        const {renderer} = this.runtime;
+        if (!renderer) return;
+
+        if (this._skinId === -1 && this._skin === null && this._drawable === -1) {
+            this._skinId = renderer.createPenSkin();
+            this._skin = renderer._allSkins[this._skinId];
+            this._drawable = renderer.createDrawable();
+            renderer.setDrawableOrder(
+                this._drawable,
+                Video.ORDER
+            );
+            renderer.updateDrawableProperties(this._drawable, {
+                skinId: this._skinId
+            });
+        }
+
+        // if we haven't already created and started a preview frame render loop, do so
+        if (!this._renderPreviewFrame) {
+            renderer.updateDrawableProperties(this._drawable, {
+                visible: true
+            });
+
+            this._renderPreviewFrame = () => {
+                if (!this._renderPreviewFrame) {
```

Comment:
Check if `this._renderPreviewFrame !== renderPreviewFrame`. `this._renderPreviewFrame` could be truthy if video was quickly toggled off and on.

Reply:
I think I'd rather guard by clearing any previous/double setTimeout's -- doing a double render is fairly inexpensive here